File: src/diffpath.lua
-- diffpath.lua
local lfs = require("lfs")

-- Function to split a PATH string into a table of directories
local function split_path(path)
    local dirs = {}
    for dir in string.gmatch(path, "([^:]+)") do
        table.insert(dirs, dir)
    end
    return dirs
end

-- Function to check if a file is executable and is either a regular file or a symlink
local function is_executable(file)
    local attrs = lfs.symlinkattributes(file)
    if not attrs or attrs.mode == "directory" then
        return false
    end
    if attrs.mode == "link" then
        return true
    end
    if attrs.mode == "file" then
        return attrs.permissions:sub(3, 3) == "x"
            or attrs.permissions:sub(6, 6) == "x"
            or attrs.permissions:sub(9, 9) == "x"
    end
    return false
end

-- Function to collect all executable filenames in a directory
local function get_executables(path)
    local execs = {}
    for _, dir in ipairs(split_path(path)) do
        if lfs.attributes(dir, "mode") == "directory" then
            for file in lfs.dir(dir) do
                if file ~= "." and file ~= ".." then
                    local filepath = dir .. "/" .. file
                    if is_executable(filepath) then
                        execs[file] = true
                    end
                end
            end
        end
    end
    return execs
end

-- Function to diff two tables of executables
local function symmetric_difference(exec1, exec2)
    local unique_files = {}
    for file in pairs(exec1) do
        if not exec2[file] then
            table.insert(unique_files, file)
        end
    end
    for file in pairs(exec2) do
        if not exec1[file] then
            table.insert(unique_files, file)
        end
    end

    table.sort(unique_files)

    return unique_files
end

-- Main function
local function main()
    if #arg < 2 then
        io.stderr:write("Usage: " .. arg[0] .. " PATH1 PATH2\n")
        os.exit(1)
    end

    local path1 = arg[1]
    local path2 = arg[2]

    local execs1 = get_executables(path1)
    local execs2 = get_executables(path2)

    local unique_files = symmetric_difference(execs1, execs2)

    for _, file in ipairs(unique_files) do
        if execs1[file] then
            print(file)
        else
            print("\t" .. file)
        end
    end
end

main()

File: src/diffpath.m
#import <Foundation/Foundation.h>
#import <dirent.h>
#import <sys/stat.h>
#import <unistd.h>

// Function to check if a file is executable without following symlinks
BOOL isExecutableAtPath(int dirfd, const char* fileName)
{
    struct stat st;
    if (fstatat(dirfd, fileName, &st, AT_SYMLINK_NOFOLLOW) == -1) {
        perror("fstatat");
        return NO;
    }
    // Check if the file is either regular or symlink and executable by user, group, or others
    return S_ISLNK(st.st_mode) || (S_ISREG(st.st_mode) && (st.st_mode & (S_IXUSR | S_IXGRP | S_IXOTH)));
}

// Function to get all executable filenames in a directory without following symlinks
NSSet<NSString*>* getExecutablesInDirectory(NSString* directory)
{
    NSMutableSet<NSString*>* executables = [NSMutableSet set];
    int dirfd = open([directory fileSystemRepresentation], O_RDONLY | O_DIRECTORY);
    if (dirfd == -1) {
        perror("open");
        return executables;
    }

    DIR* dir = fdopendir(dirfd);
    if (dir) {
        struct dirent* entry;
        while ((entry = readdir(dir)) != NULL) {
            if (entry->d_type == DT_REG || entry->d_type == DT_LNK) {
                if (isExecutableAtPath(dirfd, entry->d_name)) {
                    NSString* filename = [NSString stringWithUTF8String:entry->d_name];
                    [executables addObject:filename];
                }
            }
        }
        closedir(dir);
    } else {
        perror("fdopendir");
        close(dirfd);
    }
    close(dirfd);
    return executables;
}

// Function to collect executables from PATH
NSSet<NSString*>* collectExecutables(NSString* path)
{
    NSMutableSet<NSString*>* executables = [NSMutableSet set];
    NSArray<NSString*>* directories = [path componentsSeparatedByString:@":"];
    for (NSString* directory in directories) {
        [executables unionSet:getExecutablesInDirectory(directory)];
    }
    return executables;
}

// Main function to compare executables in two PATHs and output the difference
int main(int argc, const char* argv[])
{
    @autoreleasepool {
        if (argc != 3) {
            fprintf(stderr, "Usage: diffpath PATH1 PATH2\n");
            return 1;
        }

        NSString* path1 = [NSString stringWithUTF8String:argv[1]];
        NSString* path2 = [NSString stringWithUTF8String:argv[2]];

        NSSet<NSString*>* executables1 = collectExecutables(path1);
        NSSet<NSString*>* executables2 = collectExecutables(path2);

        NSMutableSet<NSString*>* uniqueToPath1 = [executables1 mutableCopy];
        [uniqueToPath1 minusSet:executables2];

        NSMutableSet<NSString*>* uniqueToPath2 = [executables2 mutableCopy];
        [uniqueToPath2 minusSet:executables1];

        NSMutableSet<NSString*>* symmetricDifference = [uniqueToPath1 mutableCopy];
        [symmetricDifference unionSet:uniqueToPath2];

        NSArray<NSString*>* diff = [[symmetricDifference allObjects] sortedArrayUsingSelector:@selector(compare:)];

        for (NSString* entry in diff) {
            if ([executables1 containsObject:entry]) {
                printf("%s\n", [entry UTF8String]);
            } else {
                printf("\t%s\n", [entry UTF8String]);
            }
        }
    }
    return 0;
}
